File: 451-sort-characters-by-frequency/sort-characters-by-frequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        freq = {}

        for ch in s:
            freq[ch] = freq.get(ch, 0) + 1
        
        sorted_freq = dict(sorted(freq.items(), key=lambda item: item[1], reverse=True))

        result = []
        for ch, count in sorted_freq.items():
            result.append(ch * count)

        return "".join(result)

File: 451-sort-characters-by-frequency/sort-characters-by-frequency.java
class Solution {
    public String frequencySort(String s) {
        HashMap<Character, Integer> freq = new HashMap<>();
        for (char ch : s.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }

        List<Character> chars = new ArrayList<>(freq.keySet());
        Collections.sort(chars, (a, b) -> {
            int fa = freq.get(a);
            int fb = freq.get(b);

            if (fa != fb) {
                return fb - fa;
            }

            return fa - fb;
        });

        StringBuilder result = new StringBuilder();
        for (char ch : chars) {
            int frequency = freq.get(ch);

            for (int i = 0; i < frequency; i++) {
                result.append(ch);
            }
        }

        return result.toString();
    }
}